### PR TITLE
Make chart-testing check honor values informed through `--chart-set`

### LINKED ISSUE #134: Running IBM chart ibm-object-storage-plugin-2.0.8.tgz fails with License issue
We (@mmulholla and I) were working with one of the IBMers trying to run the chart verifier. This is the feedback she gave us:

1. Expose the error so that the user knows what is wrong with the chart testing in the local cluster.
2. Honor the license value in the rendered templates. See the error below with older version of chart verifier showing the errors:
3. Add more example runs that pass in the docs so that developers can compare with their environments. 

You can get their charts here: https://github.com/IBM/charts/blob/master/repo/ibm-helm/ibm-object-storage-plugin-2.0.8.tgz

dperaza@dperaza-mac Downloads % chart-verifier verify --chart-set license=true $HOME/Downloads/ibm-object-storage-plugin-2.0.8.tgz 2>&1
Error: execution error at (ibm-object-storage-plugin/templates/flex-driver-sa.yaml:3:61): License must be accepted by setting license key to true
Deleting release 'ibm-object-storage-plugin-mng7o4fjcq'...
Error: uninstall: Release not loaded: ibm-object-storage-plugin-mng7o4fjcq: release: not found
Error deleting Helm release: Error waiting for process: exit status 1
apiversion: v1
kind: verify-report
metadata:
    tool:
        verifier-version: 1.0.0
        chart-uri: /Users/dperaza/Downloads/ibm-object-storage-plugin-2.0.8.tgz
        digest: sha256:71334727f9bf10ece92c8d79eebac55a4d78cd4b21b63f0a9555310207004591
        lastCertifiedTimestamp: "2021-05-24T16:24:00.879823-04:00"
        certifiedOpenShiftVersions: 4.7.5
    chart:
        name: ibm-object-storage-plugin
        home: ""
        sources: []
        version: 2.0.8
        description: |
            Chart for deploying ibmcloud object storage plugin.
            IBM Cloud Object Storage is persistent, highly available storage
            that you can mount to apps that run in a Kubernetes cluster by using
            the IBM Cloud Object Storage plug-in. The plug-in is a Kubernetes Flex-Volume plug-in
            that connects Cloud Object Storage buckets to pods in your cluster.
        keywords:
          - IKS
          - amd64
          - Storage
          - Commercial
          - Limited
          - Other
          - ROKS
          - ICP
        maintainers:
          - name: IBM
            email: contsto2@in.ibm.com
            url: ""
        icon: https://cache.globalcatalog.cloud.ibm.com/static/cache/2461-acb0d10a1725d783/images/uploaded/icons/object-storage.png
        apiversion: v2
        condition: ""
        tags: ""
        appversion: 2.0.8
        deprecated: false
        annotations: {}
        kubeversion: '>=1.10.1-0'
        dependencies: []
        type: ""
    overrides: ""
results:
  - check: not-contain-csi-objects
    type: Mandatory
    outcome: PASS
    reason: CSI objects do not exist
  - check: chart-testing
    type: Mandatory
    outcome: FAIL
    reason: 'Error waiting for process: exit status 1'
  - check: has-readme
    type: Mandatory
    outcome: PASS
    reason: Chart has a README
  - check: contains-test
    type: Mandatory
    outcome: PASS
    reason: Chart test files exist
  - check: contains-values-schema
    type: Mandatory
    outcome: PASS
    reason: Values schema file exist
  - check: not-contains-crds
    type: Mandatory
    outcome: PASS
    reason: Chart does not contain CRDs
  - check: images-are-certified
    type: Mandatory
    outcome: PASS
    reason: |-
        Image is Red Hat certified : ibmcom/ibmcloud-object-storage-driver@sha256:e07cf3ff8d56663fc33d19f0294a0c3f969038695a1d98f60fb762cbb6324053
        Image is Red Hat certified : ibmcom/ibmcloud-object-storage-plugin@sha256:bddac7c558cea2379724360b6a1fe8a50779d093d87047e16c0f967e2c7224a0
  - check: is-helm-v3
    type: Mandatory
    outcome: PASS
    reason: API version is V2, used in Helm 3
  - check: contains-values
    type: Mandatory
    outcome: PASS
    reason: Values file exist
  - check: has-kubeversion
    type: Mandatory
    outcome: PASS
    reason: Kubernetes version specified
  - check: helm-lint
    type: Mandatory
    outcome: PASS
    reason: Helm lint successful

### PR DESCRIPTION
**Pull request by isutton**
This change list adjusts the `chart-testing` check to propagate the
values informed by the user through `--chart-set` and friends by
merging those values into the `values.yaml` file for the test context.

- [x] Basic implementation
- [x] Make sure `--chart-set` values are propagated to `chart-testing`
      check.
- [ ] Skip `chart-testing` tests in the case the file the `KUBECONFIG`
      environment variable is referring to does not exist.
	  
Closes #134.

### CODE REVIEW

**Comment by mmulholla** [comment]
Couple of questions/comments:
1. Does this require a volume when running the docker image? If so we need to think about use cases.
2. The values file created should be deleted after we are finished with it.

**Comment by isutton** [comment]
> Couple of questions/comments:
> 1. Does this require a volume when running the docker image? If so we need to think about use cases.

It does use a bind mount when executing the container.

> 2. The values file created should be deleted after we are finished
> with it.

Indeed, am already working on it.

**Comment by isutton** [comment]
Martin Mulholland ***@***.***> writes:

> Can we include a test where both a config file and an override is
> specified. One where the override changes a value in the file and one
> where it is a new value? Also what take precedence? File or set?

Sure, we should have a test for that.

The intended behavior is the same as `helm`: `--chart-values` followed
by `--chart-set` and friends in the same order informed by the user.

**Comment by isutton** [comment]
> Can we include a test where both a config file and an override is
> specified. One where the override changes a value in the file and one
> where it is a new value? Also what take precedence? File or set?

I'd rather focus this PR in making sure the issue David reported works
and open a new issue to track missing tests to clear the path for our
partners, so I took the freedom to open #140 to track this with your
suggestions.

**Review by dperaza4dustbit — APPROVED** [review]
I was able to test locally as well. This looks like a good fix to issue 134.

**Review by dperaza4dustbit — APPROVED** [review]

File: pkg/chartverifier/checks/charttesting_test.go [hunk 1]
```diff
@@ -1,74 +1,108 @@
...
+
+	positiveTestCases := []testCase{
 		{
-			config: map[string]interface{}{
-				"namespace": "ct-test-namespace",
+			description: "providing a valid k8Project value should succeed",
+			opts: CheckOptions{
+				URI: chartUri,
+				Values: map[string]interface{}{
+					"k8Project": "default",
```
Comment (dperaza4dustbit) [inline pkg/chartverifier/checks/charttesting_test.go:67]: good

File: pkg/chartverifier/checks/charttesting_test.go [hunk 2]
```diff
@@ -1,74 +1,108 @@
...
+	}
+
+	negativeTestCases := []testCase{
 		{
-			config: map[string]interface{}{
-				"releaseLabel": "chart-verifier-app.kubernetes.io/instance",
+			description: "providing a bogus k8Project should fail",
+			opts: CheckOptions{
+				URI: chartUri,
+				Values: map[string]interface{}{
```
Comment (dperaza4dustbit) [inline pkg/chartverifier/checks/charttesting_test.go:90]: Do you also want to test not passing anything, taking default ""

File: pkg/chartverifier/checks/psql-service-0.1.7/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  charts.openshift.io/name: PSQL RedHat Demo Chart
+  charts.openshift.io/provider: RedHat
+  charts.openshift.io/supportURL: https://github.com/dperaza4dustbit/helm-chart
+apiVersion: v2
+appVersion: 10.0.0
+description: A Helm chart for a RedHat Certified PSQL
+kubeVersion: 1.20.0
+name: psql-service
+type: application
+version: 0.1.7
```
Comment (dperaza4dustbit) [inline pkg/chartverifier/checks/psql-service-0.1.7/Chart.yaml:12]: Once this PR is merged is a good sample to show Nancy and how we test it.

File: pkg/chartverifier/checks/psql-service-0.1.7/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      annotations:
+        alpha.image.policy.openshift.io/resolve-names: '*'
+      labels:
+        app: {{ .Values.k8Name }}
+        phase: test
+    spec:
+      # securityContext:
+      #   runAsUser: {{ .Values.securityContext.runAsUser }}
+      #   fsGroup: {{ .Values.podSecurityContext.fsGroup }}
+      # serviceAccount: {{ .Values.serviceAccount }}
```
Comment (dperaza4dustbit) [inline pkg/chartverifier/checks/psql-service-0.1.7/templates/deployment.yaml:24]: Lets cleanup comments on chart artifacts that don't add extra info.

File: pkg/chartverifier/checks/psql-service-0.1.7/values.schema.json [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            "pattern": "^[a-z0-9-_]+$"
+          },
+          "postgresDatabase": {
+            "type": "string",
+            "pattern": "^[a-z0-9-_]+$"
+          }
+        }
+      }     
+    }   
+  }
```
Comment (dperaza4dustbit) [inline pkg/chartverifier/checks/psql-service-0.1.7/values.schema.json:59]: New line here

**Review by dperaza4dustbit — COMMENTED** [review]
Looks good, some minor comments

**Review by mmulholla — APPROVED** [review]

**Comment by dperaza4dustbit** [comment]
My only maybe before we merge is the test case for default value of ""

**Review by dperaza4dustbit — APPROVED** [review]

**Pull request merged**